src/ui/components/HostPermissions/index.js

```diff
@@ -0,0 +1,151 @@
+/* @flow */
+import * as React from 'react';
+import { compose } from 'redux';
+
+import log from 'core/logger';
+import translate from 'core/i18n/translate';
+import type { I18nType } from 'core/types/i18n';
+import Permission from 'ui/components/Permission';
+
+/* eslint-disable no-continue */
+type Props = {|
+  i18n: I18nType,
+  permissions: Array<string>,
+|};
+
+export class HostPermissionsBase extends React.Component<Props> {
+  constructor(props: Props) {
+    const { i18n } = props;
+    super(props);
+    // These should be kept in sync with Firefox's strings for webextention
+    // host permissions which can be found in
+    // https://hg.mozilla.org/mozilla-central/raw-file/tip/browser/locales/en-US/chrome/browser/browser.properties
+    this.permissionStrings = {
+      allUrls: i18n.gettext('Access your data for all websites'),
+      wildcard: (domain) => {
+        return i18n.sprintf(
+          i18n.gettext('Access your data for sites in the %(domain)s domain'),
+          { domain }
+        );
+      },
+      tooManyWildcards: (count) => {
+        return i18n.sprintf(i18n.ngettext(
+          'Access your data in %(count)s other domain',
+          'Access your data in %(count)s other domains',
+          count), { count: i18n.formatNumber(count) }
+        );
+      },
+      oneSite: (site) => {
+        return i18n.sprintf(
+          i18n.gettext('Access your data for %(site)s'),
+          { site }
+        );
+      },
+      tooManySites: (count) => {
+        return i18n.sprintf(i18n.ngettext(
+          'Access your data on %(count)s other site',
+          'Access your data on %(count)s other sites',
+          count), { count: i18n.formatNumber(count) }
+        );
+      },
+    };
+    this.hostPermissions = [];
+  }
+
+  permissionStrings: Object;
+  hostPermissions: Array<React.Element<typeof Permission>>;
+
+  // Add individual Permission components for individual site/domain permissions.
+  formatItems(items: Array<string>, messageType: string) {
+    for (const item of items) {
+      this.hostPermissions.push(
+        <Permission
+          type="hostPermission"
+          description={this.permissionStrings[messageType](item)}
+          key={item}
+        />
+      );
+    }
+  }
+
+  // Adds Permission components for a list of host permissions. If we have 4 or
+  // fewer, display them all, otherwise display the first 3 followed by an item
+  // that says "...plus N others".
+  addHostPermissions(
+    list: Array<string>,
+    singleMessageType:string,
+    moreMessageType:string
+  ) {
+    if (list.length < 5) {
+      this.formatItems(list, singleMessageType);
+    } else {
+      this.formatItems(list.slice(0, 3), singleMessageType);
```

Comment:
Rather than calculate remaining and passing that as an argument to an element in another array as a description (which I'm just finding hard to follow), if we returned items rather than modified the component's shared array we could do:

```es6
const hostPermissions = this.formatItems(list.slice(0, 4), singleMessageType);

// If there are more than four items in the list, show the first three and replace the fourth with
// a "... plus N others" message.
if (list.length > 4) {
  hostPermissions[3] = this.permissionStrings({ message: moreMessageType, remaining: list.length - 3 });
}
```

Reply:
Yes, this is *way* better. Thanks!